### PR TITLE
Add utilities for consistent focus rings

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Related to the comment https://github.com/WordPress/openverse-frontend/issues/2049#issuecomment-1358044819 by @panchovm

## Description

This PR
- defines new Tailwind utilies to use for focus rings
- adds some stories that use these utilities
- adds visual regression tests for the 'Focus' story
- updates a few usages of each of the three utilities as an example
  - updates the `VCheckbox` component to match the WordPress component library
  - updates the search result image cell so that the white outline masks content instead of appearing outside
  - updates the search result audio cell so that the white outline masks content instead of appearing outside

This PR does not
- update every component to use these utilities
- remove other implementations of focus rings

These new utilities do not include any transitions on the shadows. This is intentional as the focus movement reacting to the <kbd>Tab</kbd> must be as fast as possible. If needed in some specific cases, it can be restored with the `.transition-shadow` utility.

## Testing Instructions

0. Check out the PR and run Storybook.
1. See the stories under Meta/Focus.
2. Add some rounded styles like `rounded-sm`, `rounded` and `rounded-full` to see how the ring behaves.
3. Read the docs for Meta/Focus.

---

0. Check out the PR and run the dev server.
1. Use the keyboard navigation to focus on the following three components:
	i. <s>Filters button in the new header (slim filled style)</s>
	ii. Checkboxes in the filters sidebar (slim transparent style with a slight mod)
	iii. Image cell in the search results (bold filled style)
    iv. Audio cell in the search results (bold filled style)
    v. Audio row in the search results (slim transparent style)

## Screenshots

image

With additional `rounded-full` style using the 'Controls' panel.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like
      `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or
      a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by obulat — CHANGES_REQUESTED** [review]
Thank you for adding this, it's really helpful!

We need to support color customization for the focus ring: for the internal header modal, the ring is yellow on a dark-charcoal background. Also, which variant should we use for the filter sidebar which has a lighter charcoal background?

**Comment by dhruvkb** [comment]
Hmm, I assumed the 2x2 matrix of focus styles shared by @panchovm in the comment was exhaustive! I'll go through the mockups and see which styles are missed, starting with the two cases you shared @obulat, thanks!

**Comment by dhruvkb** [comment]
The [checkbox component](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=28524%3A24057&t=zougb0LVMoLQiyQE-0) is fascinating when it comes to focus styles. The version in the WordPress design library is very different from our implementation.

Screenshot 2022-12-23 at 12 46 30 PM

In these, the variant seems to be slim-transparent (as the border itself gets colored on focus) but with a white offset appearing INSIDE the component (which affects the inner square's size).

**Comment by panchovm** [comment]
Thanks @obulat for mentioning the color variant. That was the only remaining point I missed. The color accents we are using are `Pink`, `Yellow`, and `White`.

The focus styles designed are based on the WordPress example you shared @dhruvkb. Although the inside border is `0.5px` thicker, being `1.5px` instead of `1px` as we have. It was changed during this year as when starting the Openverse Design Library, the inside border was `1px`.

They look like this in Figma

### WordPress Design Library



### Openverse Design Library

**Comment by dhruvkb** [comment]
@panchovm two follow up questions

- Will we be updating our design system to also use the thicker borders to match the WordPress design?

- Our current checkbox is drastically different from the WordPress design, differing not only in border thickness but also focus outline style (see below). It'll be nice if you could add some mockups of the ideal checkbox for us to refer to, if not the WordPress one.

| State | UI |
|-|-|
| On focused | Screenshot 2022-12-27 at 1 52 20 AM |
| Off focused | Screenshot 2022-12-27 at 1 53 55 AM |
| On | Screenshot 2022-12-27 at 1 54 05 AM |
| Off | Screenshot 2022-12-27 at 1 54 12 AM |

**Comment by panchovm** [comment]
To your questions @dhruvkb 

> Will we be updating our design system to also use the thicker borders to match the WordPress design?

Yes. Openverse focus styles are fully based on the WordPress one.

The checkbox, and probably a few other UI elements that I am missing, are pulled from the WordPress Design Library directly, which is why they do not exist in Openverse DL.

> It'll be nice if you could add some mockups of the ideal checkbox for us to refer to, if not the WordPress one.

To not block any code work, I can update [this diagram](https://github.com/WordPress/openverse-frontend/issues/2049#issuecomment-1358044819) with the new values: `1.5px` thickness for the inside border, and the variant we need for the accent color.

Otherwise, I need to check all components in the Design Library and update the **in progress** Figma files. Design files marked as *done* are not updated nor reviewed since the PRs related to those mockups were merged.

What do you think?

File: src/components/VCheckbox/VCheckbox.vue [hunk 1]
```diff
@@ -1,17 +1,21 @@
...
+    <!--
+    The checkbox focus style is a slight variation on the `focus-slim-tx` style.
+    Because it becomes filled when checked, it also needs the
+    `checked:focus-visible:border-white` class.
+    -->
     <input
       :id="id"
       type="checkbox"
-      class="checkbox bg-white"
+      class="checkbox h-5 w-5 flex-shrink-0 appearance-none rounded-sm bg-white me-3 focus-slim-tx checked:border-dark-charcoal checked:bg-dark-charcoal checked:focus-visible:border-white disabled:border-dark-charcoal-40 disabled:bg-dark-charcoal-10 checked:disabled:border-dark-charcoal-40 checked:disabled:bg-dark-charcoal-40"
```
Comment (obulat) [inline src/components/VCheckbox/VCheckbox.vue:11]: I see that there are some snapshot changes: the border of the checkbox in resting state was light charcoal, and now it's darker. Not sure why as the border values seem not to have changed (dark-charcoal).

File: test/storybook/visual-regression/v-checkbox.spec.ts [hunk 2]
```diff
@@ -1,27 +1,93 @@
 import { test, expect } from "@playwright/test"
 
+test.describe.configure({ mode: "parallel" })
+
```
Comment (obulat) [inline test/storybook/visual-regression/v-checkbox.spec.ts:4]: It's so nice that the snapshots are smaller now!

File: test/storybook/visual-regression/focus.spec.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  expect(await elem.screenshot()).toMatchSnapshot({ name: `${name}.png` })
+}
+
+const allSlugs = ["slim-transparent", "slim-filled", "bold-filled", "colored"]
+
+test.describe.configure({ mode: "parallel" })
+
+test.describe("Focus", () => {
+  for (const slug of allSlugs) {
+    test(`focus-unfocused-${slug}`, async ({ page }) => {
```
Comment (obulat) [inline test/storybook/visual-regression/focus.spec.ts:17]: I'm not sure we need the unfocused tests, they don't seem to bring much value.

I would, however, like to see the tests with dark-charcoal or black background, as that's been a pain point for some focus rings.

**Review by obulat — CHANGES_REQUESTED** [review]
This will be such a great dev-ex change! I've had hard times with the focus rings on several occasions.
I think there are just two things I'd like address here: the dark background visual regression tests, and the checkbox border color.

File: test/storybook/visual-regression/focus.spec.ts (hunk 3 above)
Comment (dhruvkb) [reply test/storybook/visual-regression/focus.spec.ts]: Dropped the unfocused tests and added one with a dark charcoal background.

File: test/storybook/visual-regression/v-checkbox.spec.ts (hunk 2 above)
Comment (dhruvkb) [reply test/storybook/visual-regression/v-checkbox.spec.ts:4]: Yeah, it's hard to see the diffs when the actual content is a small part of the snapshot.

**Comment by dhruvkb** [comment]
The changes to the checkbox were intentional to differentiate the different states.

| Enabled | Checked | Hovered | Focused | Border | Background | Comment |
|-|-|-|-|-|-|-|
| | | | | dark-charcoal-40 | white | Resting state |
| | | | ✅ | transparent | white | Slim transparent style |
| | | ✅ | | dark-charcoal | white | Border darkens |
| | | ✅ | ✅ | transparent | white | Focus takes precedence |
| | ✅ | | | dark-charcoal | dark-charcoal | Dark charcoal fill |
| | ✅ | | ✅ | transparent | dark-charcoal | White stroke between focus & fill |
| | ✅ | ✅ | | dark-charcoal | dark-charcoal | Same as checked + resting |
| | ✅ | ✅ | ✅ | transparent | dark-charcoal | Same as checked + focused |
| ✅ | | | | dark-charcoal-40 | dark-charcoal-10 | Background darkens |
| ✅ | | | ❌ | dark-charcoal-40 | dark-charcoal-10 | Cannot focus disabled |
| ✅ | | ❌ | | dark-charcoal-40 | dark-charcoal-10 | Cannot hover disabled |
| ✅ | | ❌ | ❌ | dark-charcoal-40 | dark-charcoal-10 | Cannot focus or hover disabled |
| ✅ | ✅ | | | dark-charcoal-40 | dark-charcoal-40 | Background matches border |
| ✅ | ✅ | | ❌ | dark-charcoal-40 | dark-charcoal-40 | Cannot focus disabled |
| ✅ | ✅ | ❌ | | dark-charcoal-40 | dark-charcoal-40 | Cannot hover disabled |
| ✅ | ✅ | ❌ | ❌ | dark-charcoal-40 | dark-charcoal-40 | Cannot focus or hover disabled |

File: src/components/VCheckbox/VCheckbox.vue (hunk 1 above)
Comment (dhruvkb) [reply src/components/VCheckbox/VCheckbox.vue]: The changes to the checkbox were intentional to differentiate the different states.

| Enabled | Checked | Hovered | Focused | Border | Background | Comment |
|-|-|-|-|-|-|-|
| | | | | dark-charcoal-40 | white | Resting state |
| | | | ✅ | transparent | white | Slim transparent style |
| | | ✅ | | dark-charcoal | white | Border darkens |
| | | ✅ | ✅ | transparent | white | Focus takes precedence |
| | ✅ | | | dark-charcoal | dark-charcoal | Dark charcoal fill |
| | ✅ | | ✅ | transparent | dark-charcoal | White stroke between focus & fill |
| | ✅ | ✅ | | dark-charcoal | dark-charcoal | Same as checked + resting |
| | ✅ | ✅ | ✅ | transparent | dark-charcoal | Same as checked + focused |
| ✅ | | | | dark-charcoal-40 | dark-charcoal-10 | Background darkens |
| ✅ | ✅ | | | dark-charcoal-40 | dark-charcoal-40 | Background matches border |

Disabled checkbox cannot be focused and shows no hover interaction.

The only changes to the checkbox were
- a lighter dark-charcoal-40 border which becomes dark-charcoal on hover
- the new focus style to match the WordPress DL

Comment (dhruvkb) [reply src/components/VCheckbox/VCheckbox.vue]: Tagging @panchovm for advice.

Comment (dhruvkb) [reply src/components/VCheckbox/VCheckbox.vue]: Updated the checkbox to remove the hover state in a4ded5f.

**Review by obulat — APPROVED** [review]
LGTM!
Thank you for explaining the checkbox state changes.

**Comment by obulat** [comment]
Note on the Playwright tests: the pages tests can be flaky. The page is rendered correctly, but when you take the snapshot, the page content is shifted to the left side. I think the easiest quick fix for this is to set `retries` to 2 in the `playwright.config.ts`.

**Review by panchovm — CHANGES_REQUESTED** [review]
### Storybook
Stories are accurate 👌 they look awesome.

### Dev
All components look excellent except for the Filters button.

In the resting state, I see the grey border during focus. In the active state, the `slim-filled` style does not mask the content, it shows pink and white borders as outside borders.

Here is a screen recording from Chrome inspector.

https://user-images.githubusercontent.com/895819/209930437-5e7dcb09-c803-4262-8b65-aac38ee6c997.mp4

**Comment by panchovm** [comment]
Forgot to mention that, in [this comment](https://github.com/WordPress/openverse-frontend/issues/2049#issuecomment-1366696216), I added an updated version of the focus styles diagram that has the color variant and the `1.5px` thickness for slim

**Comment by dhruvkb** [comment]
@panchovm in this PR I decided not to update the Button component because that was a huge change affecting lots of components. I'll handle that separately.

**Comment by dhruvkb** [comment]
Merging to unblock working on the `Button`s separately.

**Comment by panchovm** [comment]
No problem 👍 Excellent work.

**Pull request merged**